contracts/SignatureValidatorManager.sol

```diff
@@ -0,0 +1,211 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity ^0.8.18;
+import {ISafeProtocolSignatureValidator} from "./interfaces/Modules.sol";
+import {RegistryManager} from "./base/RegistryManager.sol";
+import {ISafeProtocolFunctionHandler, ISafeProtocolSignatureValidatorHooks} from "./interfaces/Modules.sol";
+import {MODULE_TYPE_SIGNATURE_VALIDATOR, MODULE_TYPE_SIGNATURE_VALIDATOR_HOOKS} from "./common/Constants.sol";
+import {ISafeAccount} from "./interfaces/Accounts.sol";
+import {ISafeProtocolSignatureValidatorManager} from "./interfaces/Manager.sol";
+import {IERC165} from "@openzeppelin/contracts/utils/introspection/IERC165.sol";
+
+/**
+ * @title SignatureValidatorManager
+ * @notice This contract facilitates signature validation. It maintains the signature validator(s) per account per domain or uses a default validation scheme based on the content of the data passed.
+ *         This contract follows the Safe{Core} Protocol specification for signature validation. For more details on specification refer to https://github.com/safe-global/safe-core-protocol-specs.
+ *         Implementaion of this contract is inspired by this pull request: https://github.com/rndlabs/safe-contracts/pull/1/files.
+ *         Expected setup to use this contract for signature validation is as follows:
+ *                  - Account enables SafeProtocolManager as a fallback handler
+ *                  - Account sets SignatureValidatorManager as a function handler for a function selector e.g. 0x1626ba7e i.e. bytes4(keccak256("isValidSignature(bytes32,bytes)")
+ * @dev SignatureValidatorManager inherits RegistryManager leading to possible state drift of Registry address with the SafeProtocolManager contract.
+ *      Do not set this contract as a fallback handler of a Safe account. Using this as a fallback handler would allow unauthorised setting of signature validator and signature validator hooks.
+ */
+
+contract SignatureValidatorManager is RegistryManager, ISafeProtocolFunctionHandler, ISafeProtocolSignatureValidatorManager {
+    constructor(address _registry, address _initialOwner) RegistryManager(_registry, _initialOwner) {}
+
+    // Signature selector 0xb5c726cb
+    bytes4 public constant SIGNATURE_VALIDATOR_SELECTOR = bytes4(keccak256("Account712Signature(bytes32,bytes32,bytes)"));
+
+    // Storage
+    /**
+     * @notice Mapping to account address => domain separator => signature validator contract
+     */
+    mapping(address => mapping(bytes32 => address)) public signatureValidators;
```

Comment:
Just FYI this will not work with ERC4337, see the comment here: https://github.com/safe-global/safe-core-protocol/issues/60